### PR TITLE
AV-73296 Add scope resource

### PR DESCRIPTION
**Pull request by PaulomeeCb**
## Jira

* AV-73296

## Description

This PR implements the addition of the scope resource and performs the create, update, import, delete, and read operations for terraform.

**NOTE**: Getting_started README is updated in the collections PR to reduce redundancy- https://github.com/couchbasecloud/terraform-provider-couchbase-capella/pull/163

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [x] This change requires a documentation update
- [ ] Documentation fix/enhancement

### How was this change tested and do you have evidence? _**(REQUIRED: Select at least 1)**_

- [x] Manually tested
- [ ] Unit tested
- [ ] Acceptance tested
- [ ] Unable to test / will not test (Please provide comments in section below)

### Testing
<details open>
  <summary>Testing</summary>
  
</details>

All validations are added to README as well.

CREATE SCOPE-
terraform plan
Screen Shot 2024-02-28 at 5 39 33 PM

terraform apply
Screen Shot 2024-02-28 at 5 40 08 PM
Screen Shot 2024-02-28 at 5 40 18 PM
created in ui-
Screen Shot 2024-02-28 at 5 43 43 PM

READ SCOPE-
 terraform output new_scope-
Screen Shot 2024-03-05 at 10 49 37 PM

Made changes in ui (remote)
Added a collection in ui to the above scope-
terraform apply
Screen Shot 2024-02-28 at 5 46 15 PM
Screen Shot 2024-02-28 at 5 46 25 PM

DELETE scope-
Changed the name of the scope in the .tfvars file-
terraform plan-
Screen Shot 2024-02-29 at 3 03 02 PM

terraform apply-
Screen Shot 2024-02-29 at 3 01 33 PM
Screen Shot 2024-02-29 at 3 01 52 PM
Screen Shot 2024-02-29 at 3 02 00 PM
In ui, old scope (terraform_scope) got deleted and created new one (new_terraform_scope) -
Screen Shot 2024-02-29 at 3 09 35 PM

READ scopes-
when adding a new scope to an existing list of scopes
terraform plan-
Scope list is given in the plan-
Screen Shot 2024-03-01 at 12 05 05 PM

terraform apply-
new scope added to scope list-
Screen Shot 2024-03-01 at 12 01 07 PM
Screen Shot 2024-03-01 at 12 01 23 PM

IMPORT -
Screen Shot 2024-03-05 at 10 42 46 PM

DESTROY-
Screen Shot 2024-03-05 at 10 47 12 PM
Screen Shot 2024-03-05 at 10 47 21 PM
Screen Shot 2024-03-05 at 10 47 32 PM

UPDATE-
changed scope name in terraform.tfvars file, 
terraform apply-
Screen Shot 2024-03-05 at 10 44 19 PM
Previous scope deleted and new scope added-
Screen Shot 2024-03-05 at 10 44 33 PM
Screen Shot 2024-03-05 at 10 44 59 PM

## Required 

Checklist:
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if required)
- [x] I have run make fmt and formatted my code

## Further comments

### CODE REVIEW

File: examples/scope/main.tf [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  required_providers {
+    couchbase-capella = {
+      source = "couchbasecloud/couchbase-capella"
+    }
+  }
+}
+
+provider "couchbase-capella" {
+  authentication_token = var.auth_token
+  host                 = var.host
```
Comment (aniket-Kumar-c) [inline examples/scope/main.tf:11]: The host is only used while testing on different environments, but it is of no use to the customer, so we can remove it

File: examples/scope/variables.tf [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+variable "cluster_id" {
+  description = "Capella Cluster ID"
+}
+
+variable "bucket_id" {
+  description = "Capella Bucket ID"
+}
+
+variable "host" {
+  description = "The Host URL of Couchbase Cloud."
```
Comment (aniket-Kumar-c) [inline examples/scope/variables.tf:23]: same as above

File: internal/resources/scope.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *ProviderSourceData, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	s.Data = data
+}
+
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:80]: We can add godoc here

File: internal/resources/scope.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+func (s *Scope) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan providerschema.Scope
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	ScopeRequest := scope_api.CreateScopeRequest{
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:90]:
```suggestion
	scopeRequest := scope_api.CreateScopeRequest{
```

File: internal/resources/scope.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+	)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating scope",
+			errorMessageWhileScopeCreation+api.ParseError(err),
+		)
+		return
+	}
+
+	diags = resp.State.Set(ctx, plan)
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:124]: I believe we should not directly set the plan in the state file. Instead, we should populate the computed field of the plan before executing the set operation, similar to how 'initializeApiKeyWithPlanAndId' works in the apikey resource

File: internal/resources/scope.go [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+	for _, apiCollection := range *scopeResp.Collections {
+		//create object
+		providerschemaCollection := providerschema.NewCollection(apiCollection)
+		collectionObj, diag := types.ObjectValueFrom(ctx, providerschema.CollectionAttributeTypes(), providerschemaCollection)
+		if diag.HasError() {
+			return nil, fmt.Errorf("collection object error"), diag
+		}
+		//print data to check apiCollection
+		data, _ := json.Marshal(apiCollection)
+		fmt.Println(string(data))
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:220]: we can remove this print statement

File: internal/resources/scope.go [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+	diags = resp.State.Set(ctx, &refreshedState)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (s Scope) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	//TODO implement me
+	panic("implement me")
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:295]: If there is no update supported, we can leave this function empty. We can refer to allowedCidrs resource for guidance.

File: internal/resources/scope.go [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+}
+
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:297]: we can add godoc here

File: internal/resources/scope.go [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+// Scope is the scope resource implementation.
+type Scope struct {
+	*providerschema.Data
+}
+
+// NewScope is a helper function to simplify the provider implementation.
+func NewScope() resource.Resource {
+	return &Scope{}
+}
+
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:46]: Can you please also add validation proof for imports?

File: internal/resources/scope.go [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+	refreshedState.Collections = collectionSet
+
+	return &refreshedState, nil, nil
+}
+
+func validateScopeNameIsSameInPlanAndState(planScopeName, stateScopeName string) bool {
+	return strings.EqualFold(planScopeName, stateScopeName)
+}
+
```
Comment (aniket-Kumar-c) [inline internal/resources/scope.go:240]: We can include Godoc here as well

File: internal/schema/scope.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+		"max_ttl": types.Int64Type,
+		"name":    types.StringType,
+		"uid":     types.StringType,
+	}
+}
+
+// NewCollection creates a new collection object.
+func NewCollection(collection scope.Collection) Collection {
+	return Collection{
+		//TODO: check nil too
```
Comment (aniket-Kumar-c) [inline internal/schema/scope.go:61]: Please ensure that we have proper nil checks before merging it into the main.

File: internal/resources/scope.go (hunk 5 above)
Comment (PaulomeeCb) [reply internal/resources/scope.go:127]: @aniket-Kumar-c but the scope doesn't return a response, also only the name is a required field, collections and uid are neither required, nor optional, they are just populated from remote. So how to create func  initializeScopeWithPlanAndName? Something like this?
```
//func initializeScopeWithPlanAndName(plan providerschema.Scope, name string) providerschema.Scope {
//	plan.Name = types.StringValue(name)
//      plan.Collections = types.SetNull() - not sure how to set the collections to null, as it is of types.Set
//	plan.Uid = types.StringNull()
//	return plan
//}
```

Comment (aniket-Kumar-c) [reply internal/resources/scope.go:127]: Yes, Before setting the computed field to the state file, we need to ensure it's null. To verify the effectiveness of this operation, intentionally introduce an error during the get operation. Afterward, perform the plan again; it should successfully read from the remote server. The plan results should indicate that only the computed field has been modified.

Comment (PaulomeeCb) [reply internal/resources/scope.go:127]: @aniket-Kumar-c I tested this scenario on the current implementation-
1. Purposely introduced changes to the computed values (scope uid, collection name, uid, maxTTL) in the state file.
2. Ran terraform plan, successfully read from the remote server and indicated that computed fields have been modified-
Screen Shot 2024-03-06 at 11 05 28 PM
Screen Shot 2024-03-06 at 11 05 40 PM
So this seems to be working even now? Is my understanding correct?🤔

**Review by aniket-Kumar-c — DISMISSED** [review]
LGTM!

**Review by SaicharanCB — DISMISSED** [review]
+1

**Review by SaicharanCB — APPROVED** [review]

**Pull request merged**